### PR TITLE
Fix build from source

### PR DESCRIPTION
**Pull request by n-riesco**
@aminya @captainsafia This PR fixes building from source.

We should add a test, but I'm not sure how to go about it.

### CODE REVIEW

**Review by aminya — CHANGES_REQUESTED** [review]
Thanks! This will help Linux users. 🎉  I am still not sure why someone would want to build from the source, but this is helpful for those cases that `node-gyp-build` acts buggy.

File: scripts/prepare.js [hunk 1]
```diff
@@ -14,28 +14,16 @@ if (process.env.npm_config_zmq_external == "true") {
...
+    return console.error("child process exited with code " + child.status);
   }
+
   const message = "Succesfully build libzmq on " + Date();
-  try {
-    fs.writeFileSync(path.join(zmqDir, "BUILD_SUCCESS"), message)
-  } catch(err) {
-    console.error(err.message);
-  };
+  fs.writeFileSync(path.join(zmqDir, "BUILD_SUCCESS"), message)
```
[the remark this answers is not in the document]
Comment (n-riesco) [reply scripts/prepare.js:26]: It's handled here: https://github.com/zeromq/zeromq.js/blob/2cf84a8cc4a63b24f883420dc1dc4264394a23d1/scripts/prepare.js#L97-L101

File: scripts/prepare.js [hunk 2]
```diff
@@ -14,28 +14,16 @@ if (process.env.npm_config_zmq_external == "true") {
...
-    });
-    if (child.stderr) {
-      return console.log(Error(child.stderr))
-    }
-    if (child.status !== 0) {
-      return console.error("child process exited with code " + child.status);
-    }
-  } catch(e) {
-    return console.error(e)
+  const child = spawnSync(scriptPath, [ZMQ, ARCH], {
```
[the remark this answers is not in the document]
Comment (n-riesco) [reply scripts/prepare.js:17]: It's handled here: https://github.com/zeromq/zeromq.js/blob/2cf84a8cc4a63b24f883420dc1dc4264394a23d1/scripts/prepare.js#L97-L101

File: scripts/prepare.js [hunk 3]
```diff
@@ -14,28 +14,16 @@ if (process.env.npm_config_zmq_external == "true") {
...
+    shell: true,
```
[the remark this answers is not in the document]
Comment (n-riesco) [reply scripts/prepare.js:18]: Needed to run the shell script `build_libzmq.sh`.

[the remark this answers is not in the document]
Comment (aminya) [reply scripts/prepare.js:18]: The `shell` option is when you want to pass shell commands directly to `spawn`. For spawning a file, it is not needed as far as I know unless you can give me the result with shell set to `false`.

[the remark this answers is not in the document]
Comment (n-riesco) [reply scripts/prepare.js:18]: This is what [the documentation for `spawnSync`](https://nodejs.org/api/child_process.html#child_process_child_process_spawnsync_command_args_options) says:

```
shell <boolean> | <string> If true, runs command inside of a shell. Uses '/bin/sh' on Unix, and process.env.ComSpec on Windows. A different shell can be specified as a string. See Shell requirements and Default Windows shell. Default: false (no shell).
```

Alternatively, please, try and set `shell` to `false`; `build_libzmq.sh` won't be run.

File: scripts/prepare.js (hunk 2 above)
Comment (aminya) [inline scripts/prepare.js:17]: If you have verified that `shell : true` is needed, then we should quote the path.
```suggestion
  const child = spawnSync(`"${scriptPath}"`, [ZMQ, ARCH], {
```

**Review by aminya — CHANGES_REQUESTED** [review]

File: scripts/prepare.js (hunk 2 above)
Comment (n-riesco) [reply scripts/prepare.js:17]: OK, I'll fix that and another issue that I've noticed (Atom removes the execution permit from all the files in `scripts`).

Comment (aminya) [reply scripts/prepare.js:17]: `apm` shouldn't change anything related to the scripts (I haven't seen such a thing before). The installation happens by `npm` underneath. The issue may be that I edited the files on Windows. If the permissions are incorrect, please edit and push them to Git.

Comment (n-riesco) [reply scripts/prepare.js]: For those of you unbelievers :stuck_out_tongue_closed_eyes: , here's what a see:

```
$ cd
$ rm -rf .atom/packages/Hydrogen/
$ npm_config_zmq_external=true apm install hydrogen
Installing hydrogen to /home/nriesco/.atom/packages ✓
$ ls -la .atom/packages/Hydrogen/node_modules/\@aminya/zeromq/scripts/
total 28
drwxr-xr-x 2 nriesco nriesco 4096 Apr 18 13:51 .
drwxr-xr-x 8 nriesco nriesco 4096 Apr 18 13:51 ..
-rw-r--r-- 1 nriesco nriesco  945 Apr 18 13:51 build_libzmq.sh
-rw-r--r-- 1 nriesco nriesco  659 Apr 18 13:51 cross_compile.sh
-rw-r--r-- 1 nriesco nriesco 1172 Apr 18 13:51 download.js
-rw-r--r-- 1 nriesco nriesco 2890 Apr 18 13:51 prepare.js
-rw-r--r-- 1 nriesco nriesco  616 Apr 18 13:51 publish_docs.sh
```

**Review by aminya — APPROVED** [review]

File: scripts/prepare.js [hunk 4]
```diff
@@ -14,28 +14,15 @@ if (process.env.npm_config_zmq_external == "true") {
...
-    });
-    if (child.stderr) {
-      return console.log(Error(child.stderr))
-    }
-    if (child.status !== 0) {
-      return console.error("child process exited with code " + child.status);
-    }
-  } catch(e) {
-    return console.error(e)
+  const child = spawnSync("/bin/sh", [scriptPath, ZMQ, ARCH], {
```
Comment (aminya) [inline scripts/prepare.js:17]: I hope this doesn't break when `scriptPath` has spaces in its names.

Comment (n-riesco) [reply scripts/prepare.js:17]: > I hope this doesn't break when scriptPath has spaces in its names.

This is one of the main differences between `execSync` (we pass a command) and `spawnSync` (we pass arguments).

I went ahead and test if `spawnSync` is working as expected. See below: It does, but other things fail:
 * [`node-gyp` doesn't work with path containing spaces](https://github.com/nodejs/node-gyp/blob/4b83c3de7300457919d53f26d96ea9ad6f6bedd8/gyp/pylib/gyp/generator/make.py#L984-L989).
 * we should quote the following assignments in `build_libzmq.sh`: https://github.com/n-riesco/zeromq.js/blob/f7a437b7453cea63c874d8771fd5cf0e12cbaf02/scripts/build_libzmq.sh#L25-L26

```
$ pwd
/home/nriesco/jupyter/github/zeromq js
$ rm -rf build/Release zmq
$ npm i

> zeromq@5.2.7 install /home/nriesco/jupyter/github/zeromq js
> node-gyp-build || npm run build:libzmq

make: Entering directory '/home/nriesco/jupyter/github/zeromq js/build'
  ACTION binding_gyp_libzmq_target_prepare_build zmq/BUILD_SUCCESS
/bin/sh: 1: js/build/Release/lib.host:/home/nriesco/jupyter/github/zeromq: not found
Building libzmq for linux
build_libzmq.sh: Building version 4.2.2 for architecture x64
/home/nriesco/jupyter/github/zeromq js/scripts/build_libzmq.sh: 22: export: js/scripts: bad variable name

[...]
```

**Pull request merged**